rack-async-http-falcon-graphql-lazy-resolve/query.rb

```diff
@@ -0,0 +1,43 @@
+require "async/http/internet"
+
+class Query < GraphQL::Schema::Object
+  field :one, String, null: false
+  field :two, String, null: false
+  field :three, String, null: false
+
+  def one
+    delay_1_data do |data|
+      data["url"]
+    end
```

Comment:
```ruby
def one
  Async { delay_1_data["url"] }
end
```

Reply:
Ahh, yes... of course. Thank you! I fixed in 8857ab82057a3e9b46302e2201405a061d89a3d9 but I'm still missing graphql-batch. The cheap memoizing I'm doing here isn't working, so `delay_2_data` gets called twice. I'm curious if you have any ideas about how to solve that? (I'd like to call the data methods once, async for sure, and use the result in multiple fields.)

Reply:
Ah, yeah, the caching thing is going to be a little trickier. The reason the current code doesn't work is that `@_delay_2_data ||= ` assignment doesn't happen until after the first request has returned. However, because it's async, it's called twice before a response gets back. That's what you want. But the second request doesn't realize there's another in progress.

You need some kind of synchronization mechanization so both calls end up waiting for the response from the shared request. You can use something like [`Async::Semaphore`](https://socketry.github.io/async/source/Async/Semaphore/index.html) but that might be a little too low level. You'll either want to wrap up some kind of cache helper or look for another high level helper in the @socketry toolkit. It's the same idea you'd use something like `@mutex.synchronize { request() }` for if you were using Ruby threads. I haven't dug too deep into the other APIs but let me know if that helps.

Reply:
Right, I see now -- thank you. I'm not sure if the semaphore is the right tool, if it's just blocking as opposed to syncing, but maybe I can get it working or find another tool that's meant for this job.

Also, separately, I came up with another idea that might work for my oddball use case. I'll see if I can work up an example for that idea as well, but it might be too funky to be generalizable. 

Anyway, thanks again for the advice -- I'll report back soon!

Reply:
I'm not sure I receive `@socketry` notifications so ping me directly to ensure I get it.

I read through this and looked at the code, the discussion here is on point and this is the correct solution. I will make a follow up PR with a few minor improvements.